File: src/lib/scoring.js
// MakanWell risk-scoring engine. Faithful to questionnaire-spec v1.0.
// All point weights live in src/data/questionnaire.json (single source of truth);
// this module only implements the algorithm over them.

import questionnaire from '../data/questionnaire.json'

export const CONDITIONS = ['hypertension', 'hyperlipidaemia', 'diabetes']

export const CONDITION_LABEL = {
  hypertension: 'Hypertension',
  hyperlipidaemia: 'Hyperlipidaemia',
  diabetes: 'Diabetes',
}

// Risk-tier display copy + badge colours (spec: Health Profile screen).
export const TIER_DISPLAY = {
  High: { label: 'Red Alert', color: '#dc2626', subtext: 'Your lifestyle signals significant risk. Act now.' },
  Moderate: { label: 'Watch Out', color: '#d97706', subtext: 'Some risk factors present. Small changes go a long way.' },
  Low: { label: 'Prevention Mode', color: '#16a34a', subtext: 'Low risk now. Keep it that way.' },
}

const zero = () => ({ hypertension: 0, hyperlipidaemia: 0, diabetes: 0 })
const addPoints = (acc, points) => {
  for (const c of CONDITIONS) acc[c] += points?.[c] ?? 0
}

export function computeBMI(heightCm, weightKg) {
  const h = Number(heightCm) / 100
  if (!h || !weightKg) return null
  return Number(weightKg) / (h * h)
}

// Find the matching band/option in an ordered band list by numeric value.
function findBand(bands, value) {
  return bands.find(
    (b) => (b.min === undefined || value >= b.min) && (b.max === undefined || value < b.max),
  )
}

/**
 * Compute the full risk score from questionnaire answers.
 * @param {object} answers — shape matches QuestionnaireAnswers in the spec.
 * @returns {{raw, normalised, tier, primaryConditions, isTie}}
 */
export function calculateRisk(answers = {}) {
  const raw = zero()

  for (const q of questionnaire.questions) {
    if (!q.scored) continue

    switch (q.type) {
      case 'single': {
        const opt = q.options.find((o) => o.value === answers[q.id])
        if (opt) addPoints(raw, opt.points)
        break
      }
      case 'multi': {
        const selected = answers[q.id] || []
        for (const v of selected) {
          const opt = q.options.find((o) => o.value === v)
          if (opt) addPoints(raw, opt.points)
        }
        break
      }
      case 'bmi': {
        const bmi = computeBMI(answers.heightCm, answers.weightKg)
        if (bmi != null) {
          const band = findBand(q.bands, bmi)
          if (band) addPoints(raw, band.points)
        }
        break
      }
      case 'waist': {
        const bands = q.thresholds[answers.gender]
        const waist = Number(answers.waistCm)
        if (bands && waist) {
          const band = findBand(bands, waist)
          if (band) addPoints(raw, band.points)
        }
        break
      }
      default:
        break
    }
  }

  // Normalise each condition to 0–100 against its own max.
  const normalised = {}
  const tier = {}
  for (const c of CONDITIONS) {
    const pct = Math.round(((raw[c] / questionnaire.maxScores[c]) * 100) * 10) / 10
    normalised[c] = pct
    tier[c] = tierForScore(pct)
  }

  // Rank; co-primary if top two are within the tie threshold.
  const sorted = CONDITIONS.map((c) => [c, normalised[c]]).sort((a, b) => b[1] - a[1])
  const isTie = sorted[0][1] - sorted[1][1] <= questionnaire.tieThreshold
  const primaryConditions = isTie ? [sorted[0][0], sorted[1][0]] : [sorted[0][0]]

  return { raw, normalised, tier, primaryConditions, isTie }
}

export function tierForScore(pct) {
  const match = questionnaire.tiers.find((t) => pct >= t.min && pct < t.max)
  return match ? match.tier : 'Low'
}

// ── Food dashboard helpers ──────────────────────────────────────────────

const RATING_PRIORITY = { safe: 0, modify: 1, avoid: 2 }

/** Most restrictive rating for a dish across one or more conditions. */
export function getWorstRating(dish, conditions) {
  return conditions.reduce((worst, cond) => {
    const rating = dish.conditions?.[cond]?.rating ?? 'safe'
    return RATING_PRIORITY[rating] > RATING_PRIORITY[worst] ? rating : worst
  }, 'safe')
}

/**
 * Sort dishes safe → modify → avoid for the given primary condition(s).
 * For multiple conditions, sorts by the worst (most restrictive) rating.
 */
export function sortDishesForConditions(dishes, conditions) {
  return [...dishes].sort(
    (a, b) => RATING_PRIORITY[getWorstRating(a, conditions)] - RATING_PRIORITY[getWorstRating(b, conditions)],
  )
}

/** Fast-track path: confirmed conditions bypass scoring entirely. */
export function fastTrack(conditions) {
  return { primaryConditions: conditions, isTie: conditions.length > 1, fastTrack: true }
}

File: server/app/scoring.py
"""Port of src/lib/scoring.js — must stay in sync with questionnaire.json."""

import json
from pathlib import Path

QUESTIONNAIRE_PATH = Path(__file__).resolve().parents[2] / "src" / "data" / "questionnaire.json"

CONDITIONS = ["hypertension", "hyperlipidaemia", "diabetes"]

with open(QUESTIONNAIRE_PATH, encoding="utf-8") as f:
    QUESTIONNAIRE = json.load(f)


def compute_bmi(height_cm, weight_kg):
    h = float(height_cm) / 100 if height_cm else 0
    if not h or not weight_kg:
        return None
    return float(weight_kg) / (h * h)


def _find_band(bands, value):
    for band in bands:
        if band.get("min") is not None and value < band["min"]:
            continue
        if band.get("max") is not None and value >= band["max"]:
            continue
        return band
    return None


def _add_points(acc, points):
    for c in CONDITIONS:
        acc[c] += (points or {}).get(c, 0)


def tier_for_score(pct):
    for t in QUESTIONNAIRE["tiers"]:
        if pct >= t["min"] and pct < t["max"]:
            return t["tier"]
    return "Low"


def calculate_risk(answers=None):
    answers = answers or {}
    raw = {c: 0 for c in CONDITIONS}

    for q in QUESTIONNAIRE["questions"]:
        if not q.get("scored"):
            continue

        qtype = q["type"]
        if qtype == "single":
            opt = next((o for o in q["options"] if o["value"] == answers.get(q["id"])), None)
            if opt:
                _add_points(raw, opt.get("points"))
        elif qtype == "multi":
            for v in answers.get(q["id"]) or []:
                opt = next((o for o in q["options"] if o["value"] == v), None)
                if opt:
                    _add_points(raw, opt.get("points"))
        elif qtype == "bmi":
            bmi = compute_bmi(answers.get("heightCm"), answers.get("weightKg"))
            if bmi is not None:
                band = _find_band(q["bands"], bmi)
                if band:
                    _add_points(raw, band.get("points"))
        elif qtype == "waist":
            bands = q["thresholds"].get(answers.get("gender"))
            waist = float(answers["waistCm"]) if answers.get("waistCm") else None
            if bands and waist:
                band = _find_band(bands, waist)
                if band:
                    _add_points(raw, band.get("points"))

    normalised = {}
    tier = {}
    for c in CONDITIONS:
        pct = round((raw[c] / QUESTIONNAIRE["maxScores"][c]) * 100, 1)
        normalised[c] = pct
        tier[c] = tier_for_score(pct)

    sorted_conds = sorted(CONDITIONS, key=lambda c: normalised[c], reverse=True)
    is_tie = normalised[sorted_conds[0]] - normalised[sorted_conds[1]] <= QUESTIONNAIRE["tieThreshold"]
    primary = [sorted_conds[0], sorted_conds[1]] if is_tie else [sorted_conds[0]]

    return {
        "raw": raw,
        "normalised": normalised,
        "tier": tier,
        "primaryConditions": primary,
        "isTie": is_tie,
    }


REQUIRED_FIELDS = [
    "age", "gender", "ethnicity", "heightCm", "weightKg", "waistCm",
    "familyHistory", "activityLevel", "friedFoodFreq", "sugaryDrinksPerDay",
    "extraSalt", "smokingStatus", "sleepQuality", "stressLevel", "symptoms",
    "hawkerFrequency", "favCategories",
]


def validate_answers(answers):
    errors = []
    for field in REQUIRED_FIELDS:
        val = answers.get(field)
        if val is None or val == "" or val == []:
            errors.append(f"Missing required field: {field}")
    return errors


def build_profile(answers):
    risk = calculate_risk(answers)
    bmi = compute_bmi(answers.get("heightCm"), answers.get("weightKg"))
    return {
        "entryPath": "ai-intake",
        "answers": answers,
        "riskScore": risk,
        "primaryConditions": risk["primaryConditions"],
        "bmi": round(bmi, 1) if bmi else None,
    }
